kernel/aster-nix/src/sched/priority_scheduler.rs

```diff
@@ -17,40 +22,226 @@ pub fn init() {
 /// are always prioritized during scheduling.
 /// Normal tasks are placed in the `normal_tasks` queue and are only
 /// scheduled for execution when there are no real-time tasks.
-struct PreemptScheduler {
+struct PreemptScheduler<T: PreemptSchedInfo> {
+    rq: Vec<SpinLock<PreemptRunQueue<T>>>,
+}
+
+impl<T: PreemptSchedInfo> PreemptScheduler<T> {
+    fn new(ncpu: u32) -> Self {
+        let mut rq = Vec::with_capacity(ncpu as usize);
+        for _ in 0..ncpu {
+            rq.push(SpinLock::new(PreemptRunQueue::new()));
+        }
+        Self { rq }
+    }
+
+    /// Selects a cpu for task to run on.
+    fn select_cpu(&self, _runnable: &Arc<T>) -> u32 {
+        // FIXME: adopt more reasonable policy once we fully enable SMP.
+        0
+    }
+}
+
+impl<T: Sync + Send + PreemptSchedInfo> Scheduler<T> for PreemptScheduler<T> {
+    fn enqueue(&self, runnable: Arc<T>, flags: EnqueueFlags) -> Option<u32> {
+        let cpu = runnable.cpu();
+        if flags == EnqueueFlags::Spawn && cpu.is_some() {
+            return None;
+        }
+        let target_cpu = if cpu.is_some() {
+            cpu.unwrap()
+        } else {
+            self.select_cpu(&runnable)
+        };
+        drop(cpu);
```

Comment:
This whole block of the code, if it is to be kept, should be written in the following way to make its intention more clear.

```rust
let target_cpu = {
    let cpu = runnable.cpu();

    // It is a logic error to spawn a task twice
    debug_assert!(flags == EnqueueFlags::Spawn && cpu.is_some());

    if cpu.is_some() {
        cpu.unwrap()
    } else {
        self.select_cpu(&runnable)
    };
};
```

This new form makes it obvious that the intention of the code block is to determine the target CPU.